### ISSUE TITLE
Option serialization regression in 2.7.2

### ISSUE THREAD

**Issue by akozhemiakin** [FasterXML/jackson-module-scala#240]
I faced with some weird regression in 2.7.2 (in 2.6.5 all is fine).

Suppose you have this kind of data structure:
```scala
trait M
case class F(label: String) extends M
case class C(m: Option[M])
```

and you have some script like this:
```scala
val mapper = new ObjectMapper() with ScalaObjectMapper
mapper.registerModule(DefaultScalaModule)

val s = mapper.writeValueAsString(C(Some(F("foo"))))
Console.println(s)
```

Until you place all of this in one sbt module all will be fine. But try to place the data structure in one dependent module and the script in the other module that depends on the first one and you will get "{m:{}}" as a result of serialization (so the type information is missed).

I can create sample project showing this behavior if you wish.

**Comment by nbauernfeind** [comment]
I do think that I might need a sample project to understand this issue better if you don't mind. I made changes to Option for 2.7.3 if you would also like to try the SNAPSHOT.

**Comment by jsw** [comment]
I've run into the same issue with 2.7.3. Here's the repro repo:

https://github.com/jsw/jackson-module-scala-issue-240

**Comment by zsxwing** [comment]
Not sure if this is related. I just put the following codes in the same file:
```Scala
package foo

import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper

@JsonTypeInfo(
use = JsonTypeInfo.Id.CLASS,
include = JsonTypeInfo.As.PROPERTY,
property = "@class")
trait Parent

case class Child(string: String) extends Parent

case class OptionWrapper(obj: Option[Parent])

object Foo extends App {
  val mapper = new ObjectMapper() with ScalaObjectMapper {}.registerModule(DefaultScalaModule)
  val json = mapper.writeValueAsString(new OptionWrapper(Some(Child("this is a string"))))
  println(json)
}
```
For 2.7.3, it outputs `{"obj":{"@class":"foo.Child"}}`, which is incorrect. The `string` field of `Child` is not in the json.

For 2.6.5, it outputs the correct result: `{"obj":{"@class":"foo.Child","string":"this is a string"}}`.

**Comment by jsw** [comment]
@nbauernfeind Any comment on this issue? I provided a bare bones repro repo above.

**Comment by progos** [comment]
I face the same in 2.7.4

**Comment by poroszd** [comment]
@brharrington's fix #271 works for me.

Anyway, I found that:
- if the trait is not in an object, serialization is broken
- if the trait is in an object and typed with `@JsonType`, serialization is broken
- if the trait is in an object, serialization works.
